### PR TITLE
Don't eagerly require Rails' minitest plugin.

### LINKED ISSUE #25046: Overwriting minitest option parsing breaks minitest plugins that add options
This is not a big issue for me, just something I noticed and that seems like a bad practice / will make it impossible for other gems to add their own minitest options ...
### Steps to reproduce

```
gem install minitest-line
ruby test/integration/seed_test.rb --line 6
```
### Expected behavior

Allow other minitest plugins to add commandline options.
### Actual behavior

```
vendor/bundle/ruby/2.2.0/gems/railties-5.0.0.rc1/lib/rails/test_unit/minitest_plugin.rb:57:in `plugin_rails_options': invalid option: --line (OptionParser::InvalidOption)
    from vendor/bundle/ruby/2.2.0/gems/minitest-5.8.4/lib/minitest.rb:191:in `block (2 levels) in process_args'
```
### System configuration

**Rails version**: 5.0.0.rc1 / Minitest 5.8.4

### LINKED ISSUE #29469: Rails 5 isn't playing well with minitest-focus
I wrote an issue with minitest-focus and they said it was somethign with Rails and I should write something up here. ANything to figure out a solution!
https://github.com/seattlerb/minitest-focus/issues/12

### Steps to reproduce
Install and use minitest-focus
Set a minitest to 'focus' with a tag before the test

### Expected behavior
Only the 'focused' test will run

### Actual behavior
All tests run

### System configuration
**Rails version**:
5.1.1

**Ruby version**:
2.3.3

**Other versions**:
minitest 5.10.2
minitest-focus 1.1.2
minitest-ar-assertions 0.1.1
minitest-reporters 1.1.14

### PR DESCRIPTION
**Pull request by kaspth**
Fixes https://github.com/rails/rails/issues/29469. Fixes https://github.com/rails/rails/issues/25046.

By making the Rails minitest behave like a standard minitest plugin
we're much more likely to not break when people use other minitest
plugins. Like minitest-focus and pride.

To do this, we need to behave like minitest: require files up front
and then perform the plugin behavior via the at_exit hook.
This also saves us a fair bit of wrangling with test file loading.

Finally, since the environment and warnings options have to be applied
as early as possible, and since minitest loads plugins at_exit, they
have to be moved to the test command.

Ideally, I'd like to backport this to both 5-1-stable and 5-0-stable. It's an extensive fix for sure though.

### CODE REVIEW

File: railties/lib/minitest/rails_plugin.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
+require "active_support/core_ext/module/attribute_accessors"
```
Comment (kaspth) [inline railties/lib/minitest/rails_plugin.rb]: Renaming this file to be "minitest/*_plugin.rb" makes minitest autoload us.

File: railties/lib/rails/test_help.rb [hunk 2]
```diff
@@ -2,7 +2,7 @@
 # so fixtures aren't loaded into that environment
 abort("Abort testing: Your Rails environment is running in production mode!") if Rails.env.production?
 
-require "rails/test_unit/minitest_plugin"
+require "minitest/rails_plugin"
```
Comment (kaspth) [inline railties/lib/rails/test_help.rb]: I should try removing this altogether now.

File: railties/lib/rails/test_unit/line_filtering.rb [hunk 3]
```diff
@@ -1,78 +1,12 @@
...
+require "rails/test_unit/runner"
 
 module Rails
   module LineFiltering # :nodoc:
     def run(reporter, options = {})
-      if options[:patterns] && options[:patterns].any? { |p| p =~ /:\d+/ }
-        options[:filter] = \
-          CompositeFilter.new(self, options[:filter], options[:patterns])
-      end
+      options[:filter] = Rails::TestUnit::Runner.compose_filter(self, options[:filter])
```
Comment (kaspth) [inline railties/lib/rails/test_unit/line_filtering.rb]: In a later PR, I'd like to follow minitest-focus'es lead to make the filters regexps. I.e. move the filter definitions into `plugin_rails_init` by resolving line numbers to a specific method name.

Needs to handle a file creating multiple Runnables though and running with multiple line numbers (test/models/post_test.rb:4:8). Tried to do it for this, but faced way too much head-desking.

File: railties/lib/rails/test_unit/runner.rb [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+          # Perform manual parsing and cleanup since option parser raises on unknown options.
+          env_index = argv.index("--environment") || argv.index("-e")
+          if env_index
+            argv.delete_at(env_index)
+            environment = argv.delete_at(env_index).strip
+          end
+          ENV["RAILS_ENV"] = environment || "test"
+
+          w_index = argv.index("--warnings") || argv.index("-w")
+          $VERBOSE = argv.delete_at(w_index) if w_index
```
Comment (kaspth) [inline railties/lib/rails/test_unit/runner.rb]: I really despise this code. At first I had defined these as Thor options in the test command, but I couldn't figure out a way to make Thor remove the options from ARGV after parsing them (that also broke the test/generators/plugin_test_runner_test.rb tests).

**Comment by kaspth** [comment]
I've been running a locally `--dev` generated app and can confirm minitest plugins like `-p` and minitest-focus both work 😊

File: railties/lib/rails/commands/test/test_command.rb [hunk 5]
```diff
@@ -1,21 +1,41 @@
 require "rails/command"
-require "rails/test_unit/minitest_plugin"
+require "rails/test_unit/runner"
 
 module Rails
   module Command
     class TestCommand < Base # :nodoc:
       no_commands do
         def help
```
Comment (y-yagi) [inline railties/lib/rails/commands/test/test_command.rb]: The test runner of the rails plugin does not use `TestCommand`.
Therefore, `TestCommand#help` will not be called and rails' help will not be displayed. 

``` 
$ rails plugin new yaffle  
$ cd yaffle 
$ <edit some files> 
$ ./bin/test --help
minitest options:
    -h, --help                       Display this help.
    -s, --seed SEED                  Sets random seed. Also via env. Eg: SEED=n rake
    -v, --verbose                    Verbose. Show progress processing files.
    -n, --name PATTERN               Filter run on /regexp/ or string.
    -e, --exclude PATTERN            Exclude /regexp/ or string from run.

Known extensions: rails, pride
    -b, --backtrace                  Show the complete backtrace
    -d, --defer-output               Output test failures and errors after the test run
    -f, --fail-fast                  Abort test run on first failure or error
    -c, --[no-]color                 Enable color in the output
    -p, --pride                      Pride. Show your testing pride!
``` 

I think that it is better to keep help displayed on `plugin_rails_options`, what about it?

File: railties/lib/rails/test_unit/runner.rb [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+require "active_support/core_ext/module/attribute_accessors"
+
+module Rails
+  module TestUnit
+    class Runner
+      mattr_reader :filters, default: []
+
+      class << self
+        def options(opts)
+          opts.on("--warnings", "-w", "Enable Ruby warnings") {}
```
Comment (kirs) [inline railties/lib/rails/test_unit/runner.rb]: Would `Display Ruby warning` be more clear option name than `Enable`?

**Comment by kirs** [comment]
I cloned the branch and tried it on a dummy app with minitest-focus and minitest/pride enabled. Works just fine 👌 
Also tried to run it against `5-0-stable` and Shopify (we have a lot of minitest plugins and edge cases), but it looks like it requires a bit more effort to backport the patch to 5.0.

Great work!

File: railties/lib/rails/commands/test/test_command.rb (hunk 5 above)
Comment (kaspth) [inline railties/lib/rails/commands/test/test_command.rb]: I noticed that but opted to separate them out. If we're running with `ruby -Itest test/models --help` we aren't running directly through the Rails test command, so I liked that we're slightly indicating that in the different help output. For instance when tests are run with `ruby` we can't support the `-e` option (iirc the app is booted once `rails/test_help` is run, so invoking the test command from there won't work either).

**Comment by kaspth** [comment]
@kirs sweet! I'll try to prepare a version of this that works on 5-0-stable, so you can try it out there 😊

**Comment by kaspth** [comment]
@kirs got a branch for 5-0-stable set up: https://github.com/rails/rails/compare/5-0-stable...test-runner-no-eager-require-5-0-stable

The branches are slightly out of sync as I haven't pushed the master test fixes there (I wanna see what the failures will be on 5-0-stable first).

File: railties/lib/rails/commands/test/test_command.rb (hunk 5 above)
Comment (y-yagi) [inline railties/lib/rails/commands/test/test_command.rb]: Oh, I see. Thanks for you explanation!

File: tools/test.rb [hunk 7]
```diff
@@ -17,5 +19,7 @@ def root
 
 ActiveSupport::TestCase.extend Rails::LineFiltering
 Rails::TestUnitReporter.executable = "bin/test"
-Minitest.run_via = :rails
+
```
Comment (y-yagi) [inline tools/test.rb]: Should we need `Rails::TestUnit::Runner.parse_options(ARGV)` here?

File: railties/lib/rails/plugin/test.rb [hunk 8]
```diff
@@ -1,7 +1,9 @@
-require "rails/test_unit/minitest_plugin"
+require "rails/test_unit/runner"
+require "rails/test_unit/reporter"
 
 Rails::TestUnitReporter.executable = "bin/test"
 
-Minitest.run_via = :rails
+Rails::TestUnit::Runner.parse_options(ARGV)
+Rails::TestUnit::Runner.load_tests(ARGV)
```
Comment (y-yagi) [inline railties/lib/rails/plugin/test.rb]: Can we use `Rails::TestUnit::Runner.run`?

**Comment by kaspth** [comment]
Failed on the new secrets test again: https://travis-ci.org/rails/rails/jobs/252119183#L1973 — I'm going to merge this since it's unrelated.

**Comment by kaspth** [comment]
5-1-stable: https://github.com/rails/rails/compare/de0d9604c4ea6b70edd96e823ae2cd397e04e49b...5b070116b0

Looked at the wrong build when I pushed to 5-0-stable then tried to fix it live. I then reverted it on 5-0-stable as I couldn't reproduce the test failures locally, then I figured it the issue was with rerun snippets on Travis since the plugin is now loaded there as well. 631c596c8d
Original failing build: https://travis-ci.org/rails/rails/builds/252123620

Think these should do it! Phew 😅

**Comment by kirs** [comment]
Since this PR was merged, any failing test in ActiveRecord/ActiveJob/ActiveSupport suite would fail in the reporter step:

```
/rails/railties/lib/rails/test_unit/reporter.rb:74:in `app_root': undefined method `root' for Rails:Module (NoMethodError)
        from /rails/railties/lib/rails/test_unit/reporter.rb:52:in `relative_path_for'
        from /rails/railties/lib/rails/test_unit/reporter.rb:70:in `format_rerun_snippet'
        from /rails/railties/lib/rails/test_unit/reporter.rb:22:in `record'
        from /rails/vendor/bundle/gems/minitest-5.3.3/lib/minitest.rb:613:in `block in record'
        from /rails/vendor/bundle/gems/minitest-5.3.3/lib/minitest.rb:612:in `each'
        from /rails/vendor/bundle/gems/minitest-5.3.3/lib/minitest.rb:612:in `record'
        from /rails/vendor/bundle/gems/minitest-5.3.3/lib/minitest.rb:293:in `run_one_method'
```

I fixed the issue in https://github.com/rails/rails/pull/29818

**Comment by kaspth** [comment]
Likewise in 519467babc and 5b070116 😄
